torch/nn/utils/spectral_norm.py

```diff
@@ -0,0 +1,112 @@
+"""
+import torch
+from torch.nn.functional import normalize
+from torch.nn.parameter import Parameter
+
+
+class SpectralNorm(object):
+
+    def __init__(self, name='weight', n_power_iterations=1, eps=1e-12):
+        self.name = name
+        self.n_power_iterations = n_power_iterations
+        self.eps = eps
+
+    def compute_weight(self, module):
+        weight = module._parameters[self.name + '_org']
+        u = module._buffers[self.name + '_u']
+        height = weight.size(0)
+        weight_mat = weight.view(height, -1)
+        for _ in range(self.n_power_iterations):
+            v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
+            u = normalize(torch.matmul(weight_mat, v), dim=0, eps=self.eps)
+
+        sigma = torch.dot(u, torch.matmul(weight_mat, v))
+        weight.data /= sigma
+        return weight, u
+
+    def remove(self, module):
+        weight = module._parameters[self.name + '_org']
+        del module._parameters[self.name]
+        del module._buffers[self.name + '_u']
+        del module._parameters[self.name + '_org']
+        module.register_parameter(self.name, weight)
+
+    def __call__(self, module, inputs):
+        weight, u = self.compute_weight(module)
+        setattr(module, self.name, weight)
+        setattr(module, self.name + '_u', u)
+
+    @staticmethod
+    def apply(module, name, n_power_iterations, eps):
+        fn = SpectralNorm(name, n_power_iterations, eps)
+        weight = module._parameters[name]
+        height = weight.size(0)
+
+        u = normalize(weight.data.new(height).normal_(0, 1), dim=0, eps=fn.eps)
+        module.register_parameter(fn.name + "_org", weight)
+        module.register_buffer(fn.name + "_u", u)
+
+        module.register_forward_pre_hook(fn)
+        return fn
+
+
+def spectral_norm(module, name='weight', n_power_iterations=1, eps=1e-12):
+    r"""Applies spectral normalization to a parameter in the given module.
+
+    .. math::
+         \mathbf{W} &= \dfrac{\mathbf{W}}{\sigma(\mathbf{W})} \\
+         \sigma(\mathbf{W}) &= \max_{\mathbf{h}: \mathbf{h} \ne 0} \dfrac{\|\mathbf{W} \mathbf{h}\|_2}{\|\mathbf{h}\|_2}
+
+    Spectral normalization stabilizes the training of discriminators (critics)
+    in GANs by rescaling the weight tensor with spectral norm :math:`\sigma` of
+    the weight matrix calculated using power iteration method. If the dimension
+    of the weight tensor is greater than 2, reshaped to 2D in power iteration method
+    to get spectral norm. This is implemented via a hook that calculates
+    spectral norm and rescales weight before every :meth:`~Module.forward`
+    call.
+
+    See https://arxiv.org/abs/1802.05957
+
+    Args:
+        module (nn.Module): containing module
+        name (str, optional): name of weight parameter
+        n_power_iterations (int, optional): number of power iterations to
+            calculate spectal norm
+        eps (float, optional): epsilon for numerical stability
```

Comment:
nit: please be specific, like `... for numerical stability in calculating norms`.

Reply:
I got it.